src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,138 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private T[] elements;
+    private int size = 0;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private void ensureCapacity() {
+        T[] newElements = (T[]) new Object[elements.length * 2];
+        System.arraycopy(elements, 0, newElements, 0, size);
+        elements = newElements;
+    }
+
+    private boolean isFull() {
+        return size == elements.length;
+    }
+
+    private boolean isIndexValid(int index) {
+        return index < size && index >= 0;
+    }
+
+    private boolean isIndexValidForAdd(int index) {
+        return index <= size && index >= 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (isFull()) {
+            ensureCapacity();
+        }
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (!isIndexValidForAdd(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You tried to add element to non exist index"
+            );
+        }
+        if (isFull()) {
+            ensureCapacity();
+        }
+        T[] newElements = (T[]) new Object[elements.length];
 
+        System.arraycopy(elements, 0, newElements, 0, index);
+        newElements[index] = value;
+        System.arraycopy(elements, index, newElements, index + 1, size - index);
+        size++;
+        elements = newElements;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
```

Comment:
The `addAll(List<T> list)` method can be optimized. Instead of calling the `add(T value)` method for each element in the list, you can ensure that the array has enough capacity to hold all the elements first, then add all the elements into the array. This way, you only need to check if the array is full once.